python/bind_data_access.h

```diff
@@ -358,6 +358,35 @@ using as_ElementArrayView = as_ElementArrayViewImpl<
     Variable, DataArray, Dataset, bucket<Variable>, bucket<DataArray>,
     bucket<Dataset>, Eigen::Vector3d, Eigen::Matrix3d, scipp::python::PyObject>;
 
+template <class T, class... Ignored>
+void bind_sizes_property(pybind11::class_<T, Ignored...> &c) {
+  using Sizes = std::unordered_map<std::string, scipp::index>;
+  std::function<Sizes(const T &self)> generator;
+  // Handle different return types from dims method
+  if constexpr (std::is_same_v<T, Dataset> || std::is_same_v<T, DatasetView>) {
+    generator = [](const T &self) {
+      Sizes out;
+      const auto &dims_ = self.dims();
+      for (const auto &[dim, index] : dims_) {
+        out[dim.name()] = index;
+      }
+      return out;
+    };
+  } else {
+    generator = [](const T &self) {
+      Sizes out;
+      const auto &dims_ = self.dims();
+      for (const auto &dim : dims_.labels()) {
+        out[dim.name()] = dims_[dim];
+      }
+      return out;
+    };
+  }
+  c.def_property_readonly("sizes", generator,
+                          "Dictionary of dimension labels (dims) to sizes.",
+                          py::return_value_policy::move);
+}
```

Comment:
Seems quite complicated. Would it be simpler to either:
a. Do this on the Python size, by just setting returning `dict(zip(dims, shape))` or
b. Return `Dimensions`, and add bindings for `Dimensions::operator[]`?

Reply:
I agree, it's not a nice amount of code to have for something so trivial. Unfortunately `Dataset` and `DatasetView` return a map of dims to indices (such that dimension order is not implied), whereas that is not the case the `DataArray` and `Variable`. I think doing this on the python side would be the best approach. The downside is splitting our property implementations between c++ bindings and monkey-patching python code.

Reply:
We have split this anyway already, e.g., `_repr_html_` and `bins` are added on the Python side.

I don't see the issue with `Dataset`, it returns an `unordered_map`, so `pybind11` should give you `operator[]` automatically, without need to bind it expicitly.